### PR TITLE
Create tests to check 'Git zero-config' flow as PR checks

### PR DESCRIPTION
**Pull request by artaleks9**
### What does this PR do?
* implements launch tests to check 'Git zero-config'  flow as PR check in Openshift Ci infrastructure -> https://github.com/openshift/release/pull/37823
* Creates Dockerfile to configure ci-operator -> https://github.com/eclipse-che/che-server/pull/495
* Creates base set of tests to check repositories of GitLab.com, Bitbucket.org, GitHub.providers which includes:
 1)  there are created test accounts with a public and private repositories with a simple devfile.yaml inside, like:
    
```
schemaVersion: 2.1.0
metadata:
  name: public-repo-wksp-testname
```
or
```
schemaVersion: 2.1.0
metadata:
  name: private-repo-wksp-testname
  ```
  
  2) a check of factory resolver by API for public/private repository git providers with NO setup PAT/OAuth
  3) launch a workspace with test git provider Url repository and check of clone a project as git flow with NO setup PAT/OAuth
    
* Adds additional file `htpasswd Provider.yaml` needed to configure environment
* Adds additional file `devworkspace-test.yaml` needed to create a test workspace:
```
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: ws-name
spec:
  started: true
  template:
    projects:
      - name: project-name
        git:
          remotes:
            origin: git-repo-url
  contributions:
    - name: che-code
      uri: https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/che-incubator/che-code/latest/devfile.yaml
      components:
        - name: che-code-runtime-description
          container:
            env:
              - name: CODE_HOST
                value: 0.0.0.0
 ```

* Upadtes 'build-pr-check.yml' GitHub action to get che-server docker image with PR changes:

eclipse-che-server-·-Quay

### Screenshot/screencast of this PR
ci-pr-checks-che-server

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/21753

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: .ci/openshift-ci/common.sh [hunk 1]
```diff
@@ -0,0 +1,238 @@
+#!/bin/bash
+#
+# Copyright (c) 2019-2021 Red Hat, Inc.
```
Comment (vinokurig) [inline .ci/openshift-ci/common.sh:3]:
```suggestion
# Copyright (c) 2023 Red Hat, Inc.
```

File: .ci/openshift-ci/test-bitbucket-flow.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+#
+# Copyright (c) 2019-2021 Red Hat, Inc.
```
Comment (vinokurig) [inline .ci/openshift-ci/test-bitbucket-flow.sh:3]:
```suggestion
# Copyright (c) 2023 Red Hat, Inc.
```

File: .ci/openshift-ci/test-github-flow.sh [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+#
+# Copyright (c) 2019-2021 Red Hat, Inc.
```
Comment (vinokurig) [inline .ci/openshift-ci/test-github-flow.sh:3]:
```suggestion
# Copyright (c) 2023 Red Hat, Inc.
```

File: .ci/openshift-ci/test-gitlab-flow.sh [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+#
+# Copyright (c) 2019-2021 Red Hat, Inc.
```
Comment (vinokurig) [inline .ci/openshift-ci/test-gitlab-flow.sh:3]:
```suggestion
# Copyright (c) 2023 Red Hat, Inc.
```

File: .ci/openshift-ci/Dockerfile [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 #
-# Copyright (c) 2019-2021 Red Hat, Inc.
+# Copyright (c) 2023 Red Hat, Inc.
```
Comment (vinokurig) [inline .ci/openshift-ci/Dockerfile:3]:
```suggestion
# Copyright (c) 2019-2023 Red Hat, Inc.
```

Comment (vinokurig) [inline .ci/openshift-ci/Dockerfile:3]: This file wasn't change, do not change the license header.

Comment (artaleks9) [reply .ci/openshift-ci/Dockerfile:3]: It seems, happened a bit confusing thing..
I recently did  PR ([Create 'Dockerfile' to configure ci-operator](https://github.com/eclipse-che/che-server/pull/495)), where added this Dockerfile, but didn't take into account Valeriy's comment about header.
So I decided to fix it here.

Comment (vinokurig) [reply .ci/openshift-ci/Dockerfile:3]: Ok, so it should be `# Copyright (c) 2023 Red Hat, Inc.`

File: .ci/openshift-ci/common.sh [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+  if oc exec -it -n ${USER_PROJECT_NAME} ${WORKSPACE_POD_NAME} -- test -f /projects/${PROJECT_NAME}/${YAML_FILE_NAME}; then
+    echo "[INFO] Project file /projects/${PROJECT_NAME}/${YAML_FILE_NAME} exists."
+  else
+    echo "[ERROR] Project file /projects/${PROJECT_NAME}/${YAML_FILE_NAME} does not exist."
+    exit 1
+  fi
+}
+
+# check a project is not cloned for private repo with no OAuth/PAT setup
+testProjectIsNotCloned() {
```
Comment (dmytro-ndp) [inline .ci/openshift-ci/common.sh:175]: This method seems to be duplicating testProjectIsCloned() method.

File: .github/workflows/build-pr-check.yml [hunk 7]
```diff
@@ -26,5 +29,25 @@ jobs:
...
+        password: ${{ secrets.QUAY_PASSWORD }}
+        registry: quay.io
     - name: Build with Maven
       run: mvn -B clean install -U -Pintegration
+    - name: Build docker images
+      if: github.event_name == 'pull_request'
+      run: ./dockerfiles/build.sh --tag:${{ env.PR_IMAGE_TAG }}
+    - name: Push docker images
+      if: github.event_name == 'pull_request'
+      run: docker push quay.io/eclipse/che-server:${{ env.PR_IMAGE_TAG }}
```
Comment (dmytro-ndp) [inline .github/workflows/build-pr-check.yml:53]: There will be dozen quay.io/eclipse/che-server:{pull_request.number} images if nobody clean up them.
Do you think there is a simple way to do it automatically?

File: .ci/openshift-ci/Dockerfile (hunk 5 above)
Comment (dmytro-ndp) [reply .ci/openshift-ci/Dockerfile:3]: Correct header: `Copyright (c) 2019-2023 Red Hat, Inc.`

**Comment by dmytro-ndp** [comment]
@artaleks9 : do you have logs of successful and failed PR check execution to share?

File: .github/workflows/build-pr-check.yml (hunk 7 above)
Comment (artaleks9) [reply .github/workflows/build-pr-check.yml:53]: Yes, will be additional image which is related to every PR-number, but I think, it shouldn't reflect on quay.io repo. This image will be single for the PR, even if the PR will have many commits - the image just is rewritten. The same practise is already exist for 'che-dashboard' repository, you can take a look - https://quay.io/repository/eclipse/che-dashboard?tab=tags.
Also pay attention to quantity images which are related  with every commit to 'main' branch in the 'che-server' quay repository (https://quay.io/repository/eclipse/che-server?tab=tags), so I guess, an image with 'pr-number' tag won't affects the overall picture.

Comment (dmytro-ndp) [reply .github/workflows/build-pr-check.yml:53]: There are 587 tags on https://quay.io/repository/eclipse/che-server?tab=tags and ~26 PR per month in this repo.
So, in a year you might expect + ~300 "pr-*" tags, which is a lot, IMHO.

**Comment by artaleks9** [comment]
* _Do you have logs of successful and failed PR check execution to share_?
**Successful Logs:** 
 -- [bitbucket.org flow](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/eclipse-che_che-server/494/pull-ci-eclipse-che-che-server-main-v12-bitbucket-flow/1649134406633263104/artifacts/bitbucket-flow/che-server-test/build-log.txt)
 -- [github.com flow ](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/eclipse-che_che-server/494/pull-ci-eclipse-che-che-server-main-v12-github-flow/1649305239200731136/artifacts/github-flow/che-server-test/build-log.txt)
 -- [gitlab.com flow](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/eclipse-che_che-server/494/pull-ci-eclipse-che-che-server-main-v12-gitlab-flow/1649134406725537792/artifacts/gitlab-flow/che-server-test/build-log.txt)
 **Failed Logs:**
 -- [gitlab.com flow](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/eclipse-che_che-server/494/pull-ci-eclipse-che-che-server-main-v12-gitlab-flow/1649045944123002880/artifacts/gitlab-flow/che-server-test/build-log.txt)
 -- [bitbucket.org flow](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/eclipse-che_che-server/494/pull-ci-eclipse-che-che-server-main-v12-bitbucket-flow/1646420253808267264)
 
 * _Which test repo is used to run the test_?
**Bitbucket.org:**
-- https://chepullreq1@bitbucket.org/chepullreq/public-repo.git
-- https://chepullreq1@bitbucket.org/chepullreq/private-repo.git
**GitHub.com:**
-- https://github.com/chepullreq1/public-repo.git
-- https://github.com/chepullreq1/private-repo.git
**Gitlab.com:**
-- https://gitlab.com/chepullreq1/public-repo.git
-- https://gitlab.com/chepullreq1/private-repo.git

* _Which factory URL is used to run the test_?
-- Here there is not factory Url like when use  from UI, is used test 'devworkspace-test.yaml' (see the description of the PR ) which has 'git remotes', where is writing the Url of test git repository.
This is enough to run test workspace and to check a clone/ not clone of a test project.

* _Will there PAT or OAuth flow (accepted, refused) scenario be covered_?
This is the end point of the task, it would be cool, if it could be done without UI. 
Honestly, I don't know could we do it without UI, but need to try.
So I can't say exactly, how it will be. We decided for the start make some simple test, on this moment it is exist:
  -- check of factory resolver by API for public/private repository git providers with NO setup PAT/OAuth
  -- launch a workspace with test git provider Url public/private repository and check of clone a project  with NO setup PAT/OAuth.

**Review by svor — APPROVED** [review]
LGTM

**Comment by dmytro-ndp** [comment]
@artaleks9: thanks for the answers!
I would suggest to express it somewhere, that this MR is introducing some kind of smoke test of factory resolver, to avoid confusing.
E.g., it could be expressed in test name, like `test-github-without-oauth-pat-setup-flow.sh`

**Review by ibuziuk — APPROVED** [review]

**Comment by ibuziuk** [comment]
@artaleks9 @dmytro-ndp folks, can we merge this PR now?

File: .ci/openshift-ci/common.sh (hunk 6 above)
Comment (artaleks9) [reply .ci/openshift-ci/common.sh]: Done.

**Comment by artaleks9** [comment]
The last updates includes:
- rename exist tests to `no-pat-oauth-flow` as proposed by @dmytro-ndp 
- refactor the method `testProjectIsCloned` as proposed by @dmytro-ndp 
- add the test template 'pat-secret.yaml'
- add new tests to check git-flow with PAT setup in `.ci/openshift-ci` of `che-server` repository
- update configuration in Openshift CI to add new tests -> [Update 'che-server' component](https://github.com/openshift/release/pull/39428)
- add new methods to implement PAT setup in the `common.sh` script :
  -- initialize an `<user-che>` namespace by `che-server` API
  -- create the PAT secret in an `<user-che>` namespace
  -- check the `git-credentials` in a running workspace

**Review by dmytro-ndp — APPROVED** [review]
New changes looks good to merge.

Thanks for the handling feedback, @artaleks9

File: .ci/openshift-ci/Dockerfile (hunk 5 above)
Comment (artaleks9) [reply .ci/openshift-ci/Dockerfile:3]: @dmytro-ndp  If the file is new, just is created, then we'll point the only current year

**Pull request merged**